From 7749c9b08cbce668cd858f42d35a2b47322a860f Mon Sep 17 00:00:00 2001
From: OpenWrt Build System <openwrt@localhost>
Date: Thu, 30 Jul 2026 08:01:12 +0800
Subject: [PATCH 2/2] qca-nss-dp: split EDMA v1 RX and TX NAPI

Layer the performance candidate on top of the separately testable EDMA v1 correctness prerequisites. Give RX and TX completions independent NAPI contexts, budgets, status checks and interrupt masks. Preserve netif_receive_skb() and do not add GRO so receive aggregation remains a separate experiment.

Based on VIKINGYFY/immortalwrt a4638cd4389.

Signed-off-by: OpenWrt Build System <openwrt@localhost>
---
 hal/dp_ops/edma_dp/edma_v1/edma_cfg.c        |   2 +-
 hal/dp_ops/edma_dp/edma_v1/edma_data_plane.c |  21 +-
 hal/dp_ops/edma_dp/edma_v1/edma_data_plane.h |  15 +-
 hal/dp_ops/edma_dp/edma_v1/edma_tx_rx.c      | 287 ++++++++++---------
 4 files changed, 177 insertions(+), 148 deletions(-)

diff --git a/hal/dp_ops/edma_dp/edma_v1/edma_cfg.c b/hal/dp_ops/edma_dp/edma_v1/edma_cfg.c
index 62563bf..36d20a6 100644
--- a/hal/dp_ops/edma_dp/edma_v1/edma_cfg.c
+++ b/hal/dp_ops/edma_dp/edma_v1/edma_cfg.c
@@ -219,7 +219,7 @@ static void edma_cleanup_txcmpl_ring_res(struct edma_hw *ehw,
 	/*
 	 * Free any buffers assigned to any descriptors
 	 */
-	edma_clean_tx(ehw, txcmpl_ring);
+	edma_clean_tx(ehw, txcmpl_ring, 0);
 
 	/*
 	 * Free TxCmpl ring descriptors
diff --git a/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.c b/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.c
index 842108f..c6a30f8 100644
--- a/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.c
+++ b/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.c
@@ -405,7 +405,8 @@ void edma_cleanup(bool is_dp_override)
 		synchronize_irq(edma_hw.misc_intr);
 		free_irq(edma_hw.misc_intr, (void *)(edma_hw.pdev));
 
-		netif_napi_del(&edma_hw.napi);
+		netif_napi_del(&edma_hw.rx_napi);
+		netif_napi_del(&edma_hw.tx_napi);
 		edma_hw.napi_added = 0;
 	}
 
@@ -449,7 +450,8 @@ static int edma_if_open(struct nss_dp_data_plane_ctx *dpc,
 	if (edma_hw.active++ != 0)
 		return NSS_DP_SUCCESS;
 
-	napi_enable(&edma_hw.napi);
+	napi_enable(&edma_hw.rx_napi);
+	napi_enable(&edma_hw.tx_napi);
 
 	/*
 	 * Enable the interrupt masks.
@@ -476,7 +478,8 @@ static int edma_if_close(struct nss_dp_data_plane_ctx *dpc)
 	/*
 	 * Disable NAPI
 	 */
-	napi_disable(&edma_hw.napi);
+	napi_disable(&edma_hw.tx_napi);
+	napi_disable(&edma_hw.rx_napi);
 	return NSS_DP_SUCCESS;
 }
 
@@ -728,7 +731,7 @@ static int edma_irq_init(void)
 	 */
 	for (i = 0; i < edma_hw.txcmpl_rings; i++) {
 		err = request_irq(edma_hw.txcmpl_intr[i],
-				  edma_handle_irq, IRQF_SHARED,
+				  edma_handle_tx_irq, IRQF_SHARED,
 				  "edma_txcmpl", (void *)edma_hw.pdev);
 		if (err) {
 			pr_debug("TXCMPL ring IRQ:%d request failed\n",
@@ -743,7 +746,7 @@ static int edma_irq_init(void)
 	 */
 	for (i = 0; i < edma_hw.rxfill_rings; i++) {
 		err = request_irq(edma_hw.rxfill_intr[i],
-				  edma_handle_irq, IRQF_SHARED,
+				  edma_handle_rx_irq, IRQF_SHARED,
 				  "edma_rxfill", (void *)edma_hw.pdev);
 		if (err) {
 			pr_debug("RXFILL ring IRQ:%d request failed\n",
@@ -758,7 +761,7 @@ static int edma_irq_init(void)
 	 */
 	for (i = 0; i < edma_hw.rxdesc_rings; i++) {
 		err = request_irq(edma_hw.rxdesc_intr[i],
-				  edma_handle_irq, IRQF_SHARED,
+				  edma_handle_rx_irq, IRQF_SHARED,
 				  "edma_rxdesc", (void *)edma_hw.pdev);
 		if (err) {
 			pr_debug("RXDESC ring IRQ:%d request failed\n",
@@ -853,12 +856,14 @@ static int edma_register_netdevice(struct net_device *netdev, uint32_t macid)
 	 * NAPI add
 	 */
 	if (!edma_hw.napi_added) {
-		netif_napi_add(netdev, &edma_hw.napi, edma_napi);
+		netif_napi_add(netdev, &edma_hw.rx_napi, edma_rx_napi);
+		netif_napi_add_tx(netdev, &edma_hw.tx_napi, edma_tx_napi);
 		/*
 		 * Register the interrupt handlers and enable interrupts
 		 */
 		if (edma_irq_init() < 0) {
-			netif_napi_del(&edma_hw.napi);
+			netif_napi_del(&edma_hw.tx_napi);
+			netif_napi_del(&edma_hw.rx_napi);
 			edma_hw.netdev_arr[macid - 1] = NULL;
 			return -EINVAL;
 		}
diff --git a/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.h b/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.h
index 2266c34..fd7b210 100644
--- a/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.h
+++ b/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.h
@@ -171,8 +171,10 @@ enum edma_tx {
  * EDMA private data structure
  */
 struct edma_hw {
-	struct napi_struct napi;
-			/* napi structure */
+	struct napi_struct rx_napi;
+			/* Rx NAPI structure */
+	struct napi_struct tx_napi;
+			/* Tx completion NAPI structure */
 	struct net_device *netdev_arr[EDMA_MAX_GMACS];
 			/* netdev for each gmac port */
 	struct device_node *device_node;
@@ -279,10 +281,13 @@ enum edma_tx edma_ring_xmit(struct edma_hw *ehw,
 			    struct sk_buff *skb,
 			    struct edma_txdesc_ring *txdesc_ring);
 uint32_t edma_clean_tx(struct edma_hw *ehw,
-			struct edma_txcmpl_ring *txcmpl_ring);
-irqreturn_t edma_handle_irq(int irq, void *ctx);
+			struct edma_txcmpl_ring *txcmpl_ring,
+			int budget);
+irqreturn_t edma_handle_rx_irq(int irq, void *ctx);
+irqreturn_t edma_handle_tx_irq(int irq, void *ctx);
 irqreturn_t edma_handle_misc_irq(int irq, void *ctx);
-int edma_napi(struct napi_struct *napi, int budget);
+int edma_rx_napi(struct napi_struct *napi, int budget);
+int edma_tx_napi(struct napi_struct *napi, int budget);
 void edma_cleanup_rings(struct edma_hw *ehw);
 int edma_hw_init(struct edma_hw *ehw);
 #endif /* __NSS_DP_EDMA_DATAPLANE__ */
diff --git a/hal/dp_ops/edma_dp/edma_v1/edma_tx_rx.c b/hal/dp_ops/edma_dp/edma_v1/edma_tx_rx.c
index 4f37f2e..13208e5 100644
--- a/hal/dp_ops/edma_dp/edma_v1/edma_tx_rx.c
+++ b/hal/dp_ops/edma_dp/edma_v1/edma_tx_rx.c
@@ -136,7 +136,8 @@ int edma_alloc_rx_buffer(struct edma_hw *ehw,
  *	Reap Tx descriptors
  */
 uint32_t edma_clean_tx(struct edma_hw *ehw,
-			struct edma_txcmpl_ring *txcmpl_ring)
+			struct edma_txcmpl_ring *txcmpl_ring,
+			int budget)
 {
 	struct platform_device *pdev = ehw->pdev;
 	struct edma_txcmpl_desc *txcmpl = NULL;
@@ -161,7 +162,8 @@ uint32_t edma_clean_tx(struct edma_hw *ehw,
 	data = edma_reg_read(EDMA_REG_TXCMPL_CONS_IDX(txcmpl_ring->id));
 	cons_idx = data & EDMA_TXCMPL_CONS_IDX_MASK;
 
-	while (cons_idx != prod_idx) {
+	while ((cons_idx != prod_idx) &&
+			(!budget || (txcmpl_consumed < budget))) {
 		txcmpl = &(((struct edma_txcmpl_desc *)
 					(txcmpl_ring->desc))[cons_idx]);
 
@@ -188,12 +190,16 @@ uint32_t edma_clean_tx(struct edma_hw *ehw,
 		}
 
 		len = skb_headlen(skb);
+
 		pr_debug("skb:%px cons_idx:%d prod_idx:%d word1:0x%x\n",
 			   skb, cons_idx, prod_idx, txcmpl->status);
 
 		dma_unmap_single(&pdev->dev, daddr,
 				 len, DMA_TO_DEVICE);
-		dev_kfree_skb_any(skb);
+		if (budget)
+			napi_consume_skb(skb, budget);
+		else
+			dev_kfree_skb_any(skb);
 
 next_txcmpl_desc:
 		if (++cons_idx == txcmpl_ring->count)
@@ -420,8 +426,6 @@ next_rx_desc:
 		work_done++;
 	}
 
-	edma_alloc_rx_buffer(ehw, rxdesc_ring->rxfill);
-
 	/*
 	 * make sure the consumer index is updated
 	 * before updating the hardware
@@ -432,90 +436,116 @@ next_rx_desc:
 }
 
 /*
- * edma_napi()
- *	EDMA NAPI handler
+ * edma_rx_napi()
+ *	Process Rx descriptors and refill rings independently from Tx.
  */
-int edma_napi(struct napi_struct *napi, int budget)
+int edma_rx_napi(struct napi_struct *napi, int budget)
 {
-	struct edma_hw *ehw = container_of(napi, struct edma_hw, napi);
-	struct edma_txcmpl_ring *txcmpl_ring = NULL;
-	struct edma_rxdesc_ring *rxdesc_ring = NULL;
-	struct edma_rxfill_ring *rxfill_ring = NULL;
-
-	struct net_device *ndev;
+	struct edma_hw *ehw = container_of(napi, struct edma_hw, rx_napi);
+	struct edma_rxdesc_ring *rxdesc_ring;
+	struct edma_rxfill_ring *rxfill_ring;
+	uint32_t status;
 	int work_done = 0;
 	int i;
 
-	for (i = 0; i < ehw->rxdesc_rings; i++) {
+	for (i = 0; i < ehw->rxdesc_rings && work_done < budget; i++) {
 		rxdesc_ring = &ehw->rxdesc_ring[i];
-		work_done += edma_clean_rx(ehw, budget, rxdesc_ring);
+		work_done += edma_clean_rx(ehw, budget - work_done,
+					   rxdesc_ring);
 	}
 
-	for (i = 0; i < ehw->txcmpl_rings; i++) {
-		txcmpl_ring = &ehw->txcmpl_ring[i];
-		work_done += edma_clean_tx(ehw, txcmpl_ring);
-	}
+	for (i = 0; i < ehw->rxfill_rings; i++)
+		edma_alloc_rx_buffer(ehw, &ehw->rxfill_ring[i]);
 
-	for (i = 0; i < ehw->rxfill_rings; i++) {
-		rxfill_ring = &ehw->rxfill_ring[i];
-		work_done += edma_alloc_rx_buffer(ehw, rxfill_ring);
-	}
+	if (work_done < budget) {
+		status = 0;
+		for (i = 0; i < ehw->rxdesc_rings; i++) {
+			rxdesc_ring = &ehw->rxdesc_ring[i];
+			status |= edma_reg_read(
+				EDMA_REG_RXDESC_INT_STAT(rxdesc_ring->id)) &
+				EDMA_RXDESC_RING_INT_STATUS_MASK;
+		}
 
-	/*
-	 * Resume netdev Tx queue
-	 */
-	/*
-	 * TODO works currently since we have a single queue.
-	 * Need to make sure we have support in place when there is
-	 * support for multiple queues
-	 */
-	for (i = 0; i < EDMA_MAX_GMACS; i++) {
-		ndev = ehw->netdev_arr[i];
-		if (!ndev)
-			continue;
+		for (i = 0; i < ehw->rxfill_rings; i++) {
+			rxfill_ring = &ehw->rxfill_ring[i];
+			status |= edma_reg_read(
+				EDMA_REG_RXFILL_INT_STAT(rxfill_ring->id)) &
+				EDMA_RXFILL_RING_INT_STATUS_MASK;
+		}
 
-		if (netif_queue_stopped(ndev) && netif_carrier_ok(ndev))
-			netif_start_queue(ndev);
+		if (status)
+			return budget;
+
+		if (napi_complete_done(napi, work_done)) {
+			for (i = 0; i < ehw->rxdesc_rings; i++) {
+				rxdesc_ring = &ehw->rxdesc_ring[i];
+				edma_reg_write(
+					EDMA_REG_RXDESC_INT_MASK(rxdesc_ring->id),
+					ehw->rxdesc_intr_mask);
+			}
+
+			for (i = 0; i < ehw->rxfill_rings; i++) {
+				rxfill_ring = &ehw->rxfill_ring[i];
+				edma_reg_write(
+					EDMA_REG_RXFILL_INT_MASK(rxfill_ring->id),
+					ehw->rxfill_intr_mask);
+			}
+		}
 	}
 
-	/*
-	 * TODO - rework and fix the budget control
-	 */
-	if (work_done < budget) {
-		/*
-		 * TODO per core NAPI
-		 */
-		napi_complete(napi);
+	return work_done;
+}
 
-		/*
-		 * Set RXDESC ring interrupt mask
-		 */
-		for (i = 0; i < ehw->rxdesc_rings; i++) {
-			rxdesc_ring = &ehw->rxdesc_ring[i];
-			edma_reg_write(
-				EDMA_REG_RXDESC_INT_MASK(rxdesc_ring->id),
-				ehw->rxdesc_intr_mask);
+/*
+ * edma_tx_napi()
+ *	Process Tx completions independently from Rx.
+ */
+int edma_tx_napi(struct napi_struct *napi, int budget)
+{
+	struct edma_hw *ehw = container_of(napi, struct edma_hw, tx_napi);
+	struct edma_txcmpl_ring *txcmpl_ring;
+	struct net_device *ndev;
+	uint32_t status;
+	int work_done = 0;
+	int i;
+
+	for (i = 0; i < ehw->txcmpl_rings && work_done < budget; i++) {
+		txcmpl_ring = &ehw->txcmpl_ring[i];
+		work_done += edma_clean_tx(ehw, txcmpl_ring,
+					   budget - work_done);
+	}
+
+	if (work_done) {
+		for (i = 0; i < EDMA_MAX_GMACS; i++) {
+			ndev = ehw->netdev_arr[i];
+			if (ndev && netif_queue_stopped(ndev) &&
+					netif_carrier_ok(ndev))
+				netif_wake_queue(ndev);
 		}
+	}
 
-		/*
-		 * Set TXCMPL ring interrupt mask
-		 */
+	if (work_done < budget) {
+		status = 0;
 		for (i = 0; i < ehw->txcmpl_rings; i++) {
 			txcmpl_ring = &ehw->txcmpl_ring[i];
-			edma_reg_write(EDMA_REG_TX_INT_MASK(txcmpl_ring->id),
-					ehw->txcmpl_intr_mask);
+			status |= edma_reg_read(
+				EDMA_REG_TX_INT_STAT(txcmpl_ring->id)) &
+				EDMA_TXCMPL_RING_INT_STATUS_MASK;
 		}
 
-		/*
-		 * Set RXFILL ring interrupt mask
-		 */
-		for (i = 0; i < ehw->rxfill_rings; i++) {
-			rxfill_ring = &ehw->rxfill_ring[i];
-			edma_reg_write(EDMA_REG_RXFILL_INT_MASK(
-						rxfill_ring->id),
-						edma_hw.rxfill_intr_mask);
+		if (status)
+			return budget;
+
+		if (napi_complete_done(napi, work_done)) {
+			for (i = 0; i < ehw->txcmpl_rings; i++) {
+				txcmpl_ring = &ehw->txcmpl_ring[i];
+				edma_reg_write(
+					EDMA_REG_TX_INT_MASK(txcmpl_ring->id),
+					ehw->txcmpl_intr_mask);
+			}
 		}
 	}
+
 	return work_done;
 }
 
@@ -702,96 +732,85 @@ irqreturn_t edma_handle_misc_irq(int irq, void *ctx)
 }
 
 /*
- * edma_handle_irq()
- *	Process IRQ and schedule napi
+ * edma_handle_tx_irq()
+ *	Mask Tx completion interrupts and schedule Tx NAPI.
  */
-irqreturn_t edma_handle_irq(int irq, void *ctx)
+irqreturn_t edma_handle_tx_irq(int irq, void *ctx)
 {
-	uint32_t reg_data = 0;
-	uint32_t rxdesc_intr_status = 0;
-	uint32_t txcmpl_intr_status = 0;
-	uint32_t rxfill_intr_status = 0;
+	struct platform_device *pdev = ctx;
+	struct edma_txcmpl_ring *txcmpl_ring;
+	struct edma_hw *ehw;
+	uint32_t status = 0;
 	int i;
-	struct edma_txcmpl_ring *txcmpl_ring = NULL;
-	struct edma_rxdesc_ring *rxdesc_ring = NULL;
-	struct edma_rxfill_ring *rxfill_ring = NULL;
-	struct edma_hw *ehw = NULL;
-	struct platform_device *pdev = (struct platform_device *)ctx;
 
 	ehw = platform_get_drvdata(pdev);
-	if (!ehw) {
-		pr_info("Unable to retrieve platrofm data");
-		return IRQ_HANDLED;
-	}
-
-	/*
-	 * Read RxDesc intr status
-	 */
-	for (i = 0; i < ehw->rxdesc_rings; i++) {
-		rxdesc_ring = &ehw->rxdesc_ring[i];
-		reg_data = edma_reg_read(
-				EDMA_REG_RXDESC_INT_STAT(rxdesc_ring->id));
-		rxdesc_intr_status |= reg_data &
-				EDMA_RXDESC_RING_INT_STATUS_MASK;
-
-		/*
-		 * Disable RxDesc intr
-		 */
-		edma_reg_write(EDMA_REG_RXDESC_INT_MASK(rxdesc_ring->id),
-			       EDMA_MASK_INT_DISABLE);
-	}
+	if (unlikely(!ehw))
+		return IRQ_NONE;
 
-	/*
-	 * Read TxCmpl intr status
-	 */
 	for (i = 0; i < ehw->txcmpl_rings; i++) {
 		txcmpl_ring = &ehw->txcmpl_ring[i];
-		reg_data = edma_reg_read(
-				EDMA_REG_TX_INT_STAT(txcmpl_ring->id));
-		txcmpl_intr_status |= reg_data &
-				EDMA_TXCMPL_RING_INT_STATUS_MASK;
+		if (!(edma_reg_read(EDMA_REG_TX_INT_STAT(txcmpl_ring->id)) &
+				EDMA_TXCMPL_RING_INT_STATUS_MASK))
+			continue;
 
-		/*
-		 * Disable TxCmpl intr
-		 */
+		status = 1;
 		edma_reg_write(EDMA_REG_TX_INT_MASK(txcmpl_ring->id),
 				EDMA_MASK_INT_DISABLE);
 	}
 
-	/*
-	 * Read RxFill intr status
-	 */
-	for (i = 0; i < ehw->rxfill_rings; i++) {
-		rxfill_ring = &ehw->rxfill_ring[i];
-		reg_data = edma_reg_read(
-				EDMA_REG_RXFILL_INT_STAT(rxfill_ring->id));
-		rxfill_intr_status |= reg_data &
-				EDMA_RXFILL_RING_INT_STATUS_MASK;
+	if (!status)
+		return IRQ_NONE;
 
-		/*
-		 * Disable RxFill intr
-		 */
-		edma_reg_write(EDMA_REG_RXFILL_INT_MASK(rxfill_ring->id),
-			       EDMA_MASK_INT_DISABLE);
+	if (likely(napi_schedule_prep(&ehw->tx_napi)))
+		__napi_schedule(&ehw->tx_napi);
 
-	}
+	return IRQ_HANDLED;
+}
 
-	if ((rxdesc_intr_status == 0) && (txcmpl_intr_status == 0) &&
-			(rxfill_intr_status == 0))
+/*
+ * edma_handle_rx_irq()
+ *	Mask Rx descriptor/fill interrupts and schedule Rx NAPI.
+ */
+irqreturn_t edma_handle_rx_irq(int irq, void *ctx)
+{
+	struct platform_device *pdev = ctx;
+	struct edma_rxdesc_ring *rxdesc_ring;
+	struct edma_rxfill_ring *rxfill_ring;
+	struct edma_hw *ehw;
+	uint32_t status = 0;
+	int i;
+
+	ehw = platform_get_drvdata(pdev);
+	if (unlikely(!ehw))
 		return IRQ_NONE;
 
 	for (i = 0; i < ehw->rxdesc_rings; i++) {
 		rxdesc_ring = &ehw->rxdesc_ring[i];
+		if (!(edma_reg_read(EDMA_REG_RXDESC_INT_STAT(rxdesc_ring->id)) &
+				EDMA_RXDESC_RING_INT_STATUS_MASK))
+			continue;
+
+		status = 1;
 		edma_reg_write(EDMA_REG_RXDESC_INT_MASK(rxdesc_ring->id),
-			       EDMA_MASK_INT_DISABLE);
+				EDMA_MASK_INT_DISABLE);
 	}
 
-	/*
-	 *TODO - per core NAPI
-	 */
-	if (rxdesc_intr_status || txcmpl_intr_status || rxfill_intr_status)
-		if (likely(napi_schedule_prep(&ehw->napi)))
-			__napi_schedule(&ehw->napi);
+	for (i = 0; i < ehw->rxfill_rings; i++) {
+		rxfill_ring = &ehw->rxfill_ring[i];
+		if (!(edma_reg_read(EDMA_REG_RXFILL_INT_STAT(rxfill_ring->id)) &
+				EDMA_RXFILL_RING_INT_STATUS_MASK))
+			continue;
+
+		status = 1;
+		edma_reg_write(EDMA_REG_RXFILL_INT_MASK(rxfill_ring->id),
+				EDMA_MASK_INT_DISABLE);
+	}
+
+	if (!status)
+		return IRQ_NONE;
+
+	if (likely(napi_schedule_prep(&ehw->rx_napi)))
+		__napi_schedule(&ehw->rx_napi);
 
 	return IRQ_HANDLED;
 }
-- 
2.39.5 (Apple Git-154)

